### PR TITLE
[SPARK-46881][CORE] Support `spark.deploy.workerSelectionPolicy`

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
### What changes were proposed in this pull request?

This PR aims to introduce a new configuration, `spark.deploy.workerSelectionPolicy`, in order to allow users to decide a
policy to assign executors on one of the assignable workers.
- `CORES_FREE_ASC`: Choose a worker with the least free cores
- `CORES_FREE_DESC`: Choose a worker with the most free cores (**default**)
- `MEMORY_FREE_ASC`: Choose a worker with the least free memory
- `MEMORY_FREE_DESC`: Choose a worker with the most free memory
- `WORKER_ID`: Choose a worker with the smallest worker id

### Why are the changes needed?

This is very helpful for users to manage Spark Standalone cluster dynamically based on their workload and underlying cluster management policy.

### Does this PR introduce _any_ user-facing change?

No. The default policy, `CORES_FREE_DESC`, follows the existing behavior.

### How was this patch tested?

Pass the CIs with the newly added test cases.

### Was this patch authored or co-authored using generative AI tooling?

No.

### CODE REVIEW

**Comment by dongjoon-hyun** [comment]
Could you review this when you have some time, @viirya ?

File: core/src/main/scala/org/apache/spark/deploy/master/Master.scala [hunk 1]
```diff
@@ -824,9 +827,15 @@ private[deploy] class Master(
         // If the cores left is less than the coresPerExecutor,the cores left will not be allocated
         if (app.coresLeft >= coresPerExecutor) {
           // Filter out workers that don't have enough resources to launch an executor
-          val usableWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
+          val aliveWorkers = workers.toArray.filter(_.state == WorkerState.ALIVE)
             .filter(canLaunchExecutor(_, resourceDesc))
-            .sortBy(_.coresFree).reverse
+          val usableWorkers = workerSelectionPolicy match {
+            case CORES_FREE_ASC => aliveWorkers.sortBy(w => (w.coresFree, w.id))
+            case CORES_FREE_DESC => aliveWorkers.sortBy(w => (w.coresFree, w.id)).reverse
```
Comment (viirya) [inline core/src/main/scala/org/apache/spark/deploy/master/Master.scala:834]: Hmm, this seems to sort workers by `(free cores, id)`? Shouldn't we sort them by free cores only?

Comment (dongjoon-hyun) [reply core/src/main/scala/org/apache/spark/deploy/master/Master.scala:834]: Ya, the ID is added to be deterministic. Previously, it's based on HashSet and indeterministic among `coresFree`.

**Review by viirya — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
Thank you, @viirya !

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 4.

File: core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala [hunk 2]
```diff
@@ -661,6 +662,52 @@ class MasterSuite extends SparkFunSuite
...
+    (CORES_FREE_DESC, false, List("10005")),
+    (MEMORY_FREE_ASC, true, List("10001", "10005")),
+    (MEMORY_FREE_ASC, false, List("10001")),
+    (MEMORY_FREE_DESC, true, List("10002", "10003")),
+    (MEMORY_FREE_DESC, false, Seq("10002")),
+    (WORKER_ID, true, Seq("10001", "10002")),
+    (WORKER_ID, false, Seq("10001")))
+
+  workerSelectionPolicyTestCases.foreach { case (policy, spreadOut, expected) =>
+    test(s"SPARK-46881: scheduling with workerSelectionPolicy - $policy ($spreadOut)") {
```
Comment (HyukjinKwon) [inline core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala:679]: Hm, this test seems flaky:

```
[info] - SPARK-46881: scheduling with workerSelectionPolicy - CORES_FREE_DESC (false) *** FAILED *** (178 milliseconds)
[info]   List("10004") did not equal List("10005") (MasterSuite.scala:728)
[info]   Analysis:
[info]   List(0: "10004" -> "10005")
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions.newAssertionFailedException(Assertions.scala:472)
[info]   at org.scalatest.Assertions.newAssertionFailedException$(Assertions.scala:471)
[info]   at org.scalatest.Assertions$.newAssertionFailedException(Assertions.scala:1231)
[info]   at org.scalatest.Assertions$AssertionsHelper.macroAssert(Assertions.scala:1295)
[info]   at org.apache.spark.deploy.master.MasterSuite.$anonfun$new$70(MasterSuite.scala:728)
[info]   at org.scalatest.enablers.Timed$$anon$1.timeoutAfter(Timed.scala:127)
[info]   at org.scalatest.concurrent.TimeLimits$.failAfterImpl(TimeLimits.scala:282)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter(TimeLimits.scala:231)
[info]   at org.scalatest.concurrent.TimeLimits.failAfter$(TimeLimits.scala:230)
[info]   at org.apache.spark.SparkFunSuite.failAfter(SparkFunSuite.scala:69)
[info]   at org.apache.spark.SparkFunSuite.$anonfun$test$2(SparkFunSuite.scala:155)
[info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
[info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
[info]   at org.scalatest.OutcomeOf$.outcomeOf(OutcomeOf.scala:104)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:22)
[info]   at org.scalatest.Transformer.apply(Transformer.scala:20)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike$$anon$1.apply(AnyFunSuiteLike.scala:226)
[info]   at org.apache.spark.SparkFunSuite.withFixture(SparkFunSuite.scala:227)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.invokeWithFixture$1(AnyFunSuiteLike.scala:224)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.$anonfun$runTest$1(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.SuperEngine.runTestImpl(Engine.scala:306)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest(AnyFunSuiteLike.scala:236)
[info]   at org.scalatest.funsuite.AnyFunSuiteLike.runTest$(AnyFunSuiteLike.scala:218)
[info]   at org.apache.spark.SparkFunSuite.org$scalatest$BeforeAndAfterEach$$super$runTest(SparkFunSuite.scala:69)
[info]   at org.scalatest.BeforeAndAfterEach.runTest(BeforeAndAfterEach.scala:234)
[info]   at org.scalatest.BeforeAndAfterEach.runTest$(BeforeAndAfterEach.scala:227)
```

https://github.com/apache/spark/actions/runs/8042308713/job/21962794853#step:10:17224

Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala:679]: Thanks. Ya, let me take a look at.

Comment (dongjoon-hyun) [reply core/src/test/scala/org/apache/spark/deploy/master/MasterSuite.scala:679]: To @HyukjinKwon , here is the follow-up.
- #45274

**Pull request closed without merging**